### DISCUSSION TITLE
How to produce a Protobuf serialised to byteArray

### DISCUSSION THREAD

**Discussion by maxdidato** [provectus/kafka-ui#2944] [Q&A] [answered]
Hi, 
I am trying to figure out how to produce a byte array representing the serialisation to protobuf of a java object. 
I have tried by transforming the byte array to a string UTF-8 encoded and send it via the UI. 
I have also tried to use the API to send the payload specifying the valueSerde (ByteArraySerializer). 

In both cases the consumer will try to deserialise the payload using the encoded string version of the byte array and it fails. 
If I use the Java Kafka producer and produce the byte array directly if works of course. 

Any suggestion here?
Thanks

**Comment by iliax** [comment, marked as answer]
Hello @maxdidato 
here some options for you:
1. if you are using SchemaRegistry with protobuf schema - then just configure SR for your cluster. (example is [here](https://github.com/provectus/kafka-ui/blob/master/documentation/compose/kafka-ui.yaml))

2. If you **not** using SR and just want to produce messages with specified protobuf schema - you should tell kafka-ui where that schema can be found (example is [here](https://github.com/provectus/kafka-ui/blob/master/documentation/guides/Protobuf.md)). After that you will see "ProtobufFile" serde in serde selection field. After that you just enter [json representation](https://developers.google.com/protocol-buffers/docs/proto3#json) of your msg, it will be serialized and sent to kafka. Using this approach messages should be also viewable in "Messages" tab.

3. If you need a byte-exact representation of binary data without an ability to view its content - you can serialize proto message to byte array in you app code, transform to Base64 string and use "Base64" while producing message in kafka-ui.

**Comment by maxdidato** [comment]
Hello @iliax , thanks for the response. 
My requirement was actually just to produce a Base64 encoded version of the serialised object. 
The documentation is a little bit lacking so I figured it out by looking at the code. I just needed to use the Base64 serialiizer. 
I have also noticed that in master it is already available a dropbox in the producer window to let the user decide what serialiser to use for that specific message. 
At the moment I am using that and it works perfectly.
Thanks

**Reply by Haarolean** [reply]
Glad it works!
